### ISSUE TITLE
External: Dynamic fails on netcoreapp2.0

### ISSUE THREAD

**Issue by maumar** [aspnet/EntityFramework#8962]
failing tests:

- [ ] Select_GroupBy 
- [ ] Select_GroupBy_SelectMany
- [ ] Select_nested_collection_count_using_DTO
- [ ] Select_DTO_distinct_translated_to_server
- [ ] Select_DTO_constructor_distinct_translated_to_server
- [ ] Select_DTO_with_member_init_distinct_translated_to_server
- [ ] Select_DTO_with_member_init_distinct_in_subquery_used_in_projection_translated_to_server

all failing with the same underlying issue:

```
System.ArgumentOutOfRangeException : Length cannot be less than zero.
Parameter name: length
Stack Trace:
	at System.String.Substring(Int32 startIndex, Int32 length)
	at Microsoft.CSharp.RuntimeBinder.Syntax.NameTable.Add(String key, Int32 length)
	at Microsoft.CSharp.RuntimeBinder.Syntax.NameManager.Add(String key, Int32 length)
	at Microsoft.CSharp.RuntimeBinder.SymbolTable.GetName(Type type)
	at Microsoft.CSharp.RuntimeBinder.SymbolTable.LoadSymbolsFromType(Type originalType)
	at Microsoft.CSharp.RuntimeBinder.SymbolTable.GetConstructedType(Type type, AggregateSymbol agg)
	at Microsoft.CSharp.RuntimeBinder.SymbolTable.LoadSymbolsFromType(Type originalType)
	at Microsoft.CSharp.RuntimeBinder.RuntimeBinder.GetArgumentType(ICSharpBinder p, CSharpArgumentInfo argInfo, Expression param, DynamicMetaObject arg, Int32 index)
	at Microsoft.CSharp.RuntimeBinder.RuntimeBinder.CreateArgumentArray(ICSharpBinder payload, Expression[] parameters, DynamicMetaObject[] args)
	at Microsoft.CSharp.RuntimeBinder.RuntimeBinder.BindCore(ICSharpBinder payload, Expression[] parameters, DynamicMetaObject[] args, DynamicMetaObject& deferredBinding)
	at Microsoft.CSharp.RuntimeBinder.RuntimeBinder.Bind(DynamicMetaObjectBinder payload, Expression[] parameters, DynamicMetaObject[] args, DynamicMetaObject& deferredBinding)
	at Microsoft.CSharp.RuntimeBinder.BinderHelper.Bind(DynamicMetaObjectBinder action, RuntimeBinder binder, DynamicMetaObject[] args, IEnumerable`1 arginfos, DynamicMetaObject onBindingError)
	at Microsoft.CSharp.RuntimeBinder.CSharpConvertBinder.FallbackConvert(DynamicMetaObject target, DynamicMetaObject errorSuggestion)
	at System.Dynamic.DynamicMetaObject.BindConvert(ConvertBinder binder)
	at System.Dynamic.ConvertBinder.Bind(DynamicMetaObject target, DynamicMetaObject[] args)
	at System.Dynamic.DynamicMetaObjectBinder.Bind(Object[] args, ReadOnlyCollection`1 parameters, LabelTarget returnLabel)
	at System.Runtime.CompilerServices.CallSiteBinder.BindCore[T](CallSite`1 site, Object[] args)
	at System.Dynamic.UpdateDelegates.UpdateAndExecute1[T0,TRet](CallSite site, T0 arg0)
	at System.Linq.EnumerableSorter`2.ComputeKeys(TElement[] elements, Int32 count)
	at System.Linq.EnumerableSorter`1.ComputeMap(TElement[] elements, Int32 count)
	at System.Linq.EnumerableSorter`1.Sort(TElement[] elements, Int32 count)
	at System.Linq.OrderedEnumerable`1.ToList()
	at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
```

**Comment by ajcvickers** [comment]
Spend a few hours trying to get a repro of the .NET Core difference.

**Comment by maumar** [comment]
I was able to narrow it down to the following test method in our complex navs tests:

```
        public class MyEntity
        {
            public int Id { get; set; }
            public string Name { get; set; }
        }

        [ConditionalFact]
        public virtual void Netcoreapp_bug()
        {
            Func<IQueryable<MyEntity>, IQueryable<object>> func =
                es =>
                    es.Select(e => new MyOuterDto
                    {
                        Id = e.Id,
                        Name = e.Name
                    });

            Action<dynamic, dynamic> elementAsserter = (e, a) => Console.WriteLine(e.Id);

            var entities = new List<MyEntity>
            {
                new MyEntity {  Id = 1, Name = "1" },
                new MyEntity {  Id = 2, Name = "2" },
                new MyEntity {  Id = 3, Name = "3" }
            }.AsQueryable();

            var query = func(entities).ToArray();

            for (var i = 0; i < query.Length; i++)
            {
                elementAsserter(query[i], query[i]);
            }
        }
```

The method doesn't contain any reference to EF code, however I wasn't able to reproduce it in a standalone netcoreapp project, so there is something in a way our project is setup that's causing the issue. @smitpatel volunteered to take a stab at it, as he knows the infrastructure well.

**Comment by smitpatel** [comment]
>as he knows the infrastructure well.

😨

**Comment by maumar** [comment]
filed https://github.com/dotnet/coreclr/issues/12529

**Comment by maumar** [comment]
addressed in c781e01b4edcd75f3c4016e6d4ca1c87cde59940

**Comment by JonHanna** [comment]
> Perhaps there is something in a way our project is setup that's causing the issue t surface.

Yes, a generic type within which `MyEntity` is nested, making it generic. The bug is with the handling of types which are generic because they are nested in a generic type but which do not themselves have parameter types.

I'm afraid when I reduced some allocations in `SymbolTable.GetName` I introduced an assumption that all generic types have backticks in their names that doesn't hold for such nested types.
